staging/src/k8s.io/cloud-provider/controllers/route/route_controller.go

```diff
@@ -108,18 +184,66 @@ func (rc *RouteController) Run(ctx context.Context, syncPeriod time.Duration, co
 		return
 	}
 
-	// TODO: If we do just the full Resync every 5 minutes (default value)
-	// that means that we may wait up to 5 minutes before even starting
-	// creating a route for it. This is bad.
-	// We should have a watch on node and if we observe a new node (with CIDR?)
-	// trigger reconciliation for that node.
-	go wait.NonSlidingUntil(func() {
+	if utilfeature.DefaultFeatureGate.Enabled(features.CloudControllerManagerWatchBasedRoutesReconciliation) {
+		// Process any node events that may change the routes.
+		// It does not make sense to add concurrency here, as the routes
+		// controller always processes the whole cluster.
+		go wait.UntilWithContext(ctx, rc.runWorker, time.Second)
+	} else {
+		go wait.NonSlidingUntil(func() {
+			if err := rc.reconcileNodeRoutes(ctx); err != nil {
+				klog.Errorf("Couldn't reconcile node routes: %v", err)
+			}
+		}, syncPeriod, ctx.Done())
+	}
+
+	<-ctx.Done()
+}
+
+func (rc *RouteController) runWorker(ctx context.Context) {
+	for rc.processNextWorkItem(ctx) {
+	}
```

Comment:
should something be in here?

Reply:
With Elmiko here. Pushing the processing into the loop check seems wrong. It also leaves an empty loop which tends to catch the eye and at least looks like a mistake.